### ISSUE TITLE
Bad results with -split_input

### ISSUE THREAD

**Issue by chamlin** [marklogic/marklogic-contentpump#57]
I checked out mlcp and built via the instructions at https://github.com/marklogic/marklogic-contentpump

Then unzipped the mlcp bin deliverable and 
``` bash
$ marklogic-contentpump/mlcp/deliverable/mlcp-9.0-EA3/bin/mlcp.sh version
```
gave
``` bash
chamlin@MacPro-3445:bug$ /Users/chamlin/tickets/18184-lds-mlcp/marklogic-contentpump/mlcp/deliverable/mlcp-9.0-EA3/bin/mlcp.sh version
ContentPump version: 9.0-EA3
Java version: 1.8.0_51
Hadoop version: 2.6.0
Supported MarkLogic versions: 6.0 - 9.0
```
Next run
``` bash
$ marklogic-contentpump/mlcp/deliverable/mlcp-9.0-EA3/bin/mlcp.sh -options_file test-quotes.txt
17/02/02 21:29:11 INFO contentpump.LocalJobRunner: Content type: XML
17/02/02 21:29:12 INFO contentpump.ContentPump: Job name: local_755095747_1
17/02/02 21:29:12 INFO contentpump.FileAndDirectoryInputFormat: Total input paths to process : 1
17/02/02 21:29:13 INFO contentpump.LocalJobRunner:  completed 100%
17/02/02 21:29:13 INFO contentpump.LocalJobRunner: com.marklogic.mapreduce.MarkLogicCounter: 
17/02/02 21:29:13 INFO contentpump.LocalJobRunner: INPUT_RECORDS: 900
17/02/02 21:29:13 INFO contentpump.LocalJobRunner: OUTPUT_RECORDS: 900
17/02/02 21:29:13 INFO contentpump.LocalJobRunner: OUTPUT_RECORDS_COMMITTED: 900
17/02/02 21:29:13 INFO contentpump.LocalJobRunner: OUTPUT_RECORDS_FAILED: 0
17/02/02 21:29:13 INFO contentpump.LocalJobRunner: Total execution time: 0 sec
```
This reports 900 records and I see 900 in the database.  All is well.

But, if I uncomment the lines
```
#-split_input 
#true
#--max_split_size
#1000
```
and rerun I get
``` bash
marklogic-contentpump/mlcp/deliverable/mlcp-9.0-EA3/bin/mlcp.sh -options_file test-quotes.txt
17/02/02 21:45:30 INFO contentpump.LocalJobRunner: Content type: XML
17/02/02 21:45:30 INFO contentpump.ContentPump: Job name: local_2038694039_1
17/02/02 21:45:30 INFO contentpump.FileAndDirectoryInputFormat: Total input paths to process : 1
17/02/02 21:45:30 INFO contentpump.DelimitedTextInputFormat: 72 DelimitedSplits generated
17/02/02 21:45:31 ERROR contentpump.LocalJobRunner: Error running task: 
java.lang.RuntimeException: java.io.IOException: (line 2) invalid char between encapsulated token and delimiter
	at org.apache.commons.csv.CSVParser$1.getNextRecord(CSVParser.java:442)
	at org.apache.commons.csv.CSVParser$1.hasNext(CSVParser.java:452)
	at com.marklogic.contentpump.SplitDelimitedTextReader.initParser(SplitDelimitedTextReader.java:205)
	at com.marklogic.contentpump.SplitDelimitedTextReader.initialize(SplitDelimitedTextReader.java:62)
	at com.marklogic.contentpump.LocalJobRunner$TrackingRecordReader.initialize(LocalJobRunner.java:439)
	at com.marklogic.contentpump.LocalJobRunner$LocalMapTask.call(LocalJobRunner.java:373)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: (line 2) invalid char between encapsulated token and delimiter
	at org.apache.commons.csv.Lexer.parseEncapsulatedToken(Lexer.java:275)
	at org.apache.commons.csv.Lexer.nextToken(Lexer.java:152)
	at org.apache.commons.csv.CSVParser.nextRecord(CSVParser.java:498)
	at org.apache.commons.csv.CSVParser$1.getNextRecord(CSVParser.java:439)
	... 9 more
17/02/02 21:45:31 INFO contentpump.LocalJobRunner:  completed 1%
17/02/02 21:45:31 INFO contentpump.LocalJobRunner: com.marklogic.mapreduce.MarkLogicCounter: 
17/02/02 21:45:31 INFO contentpump.LocalJobRunner: INPUT_RECORDS: 910
17/02/02 21:45:31 INFO contentpump.LocalJobRunner: OUTPUT_RECORDS: 910
17/02/02 21:45:31 INFO contentpump.LocalJobRunner: OUTPUT_RECORDS_COMMITTED: 910
17/02/02 21:45:31 INFO contentpump.LocalJobRunner: OUTPUT_RECORDS_FAILED: 0
17/02/02 21:45:31 INFO contentpump.LocalJobRunner: Total execution time: 1 sec
```
and ML shows 888 in the db after the run.  So the count shows more, and the db shows less.

I did a little debugging in SplitDelimitedTextReader.java, to show the start/end of the splits.  It appears the file is OK, but when the split falls (after moving back one) on the second quote in a line, you get this parse error.  It turns out that parserIterator.hasNext() actually reads a record, and the parse fails.  The 888 is explained by the records thrown away when the split has the error.  I have no idea why the input records are even higher when the split is tossed, certainly there weren't 910 records commited.  No errors showed in the log file.

Running the other file as
``` bash
marklogic-contentpump/mlcp/deliverable/mlcp-9.0-EA3/bin/mlcp.sh -options_file test-jp.txt
```
gives
``` bash
17/02/02 21:53:03 INFO contentpump.LocalJobRunner: Content type: XML
17/02/02 21:53:03 INFO contentpump.ContentPump: Job name: local_367057373_1
17/02/02 21:53:03 INFO contentpump.FileAndDirectoryInputFormat: Total input paths to process : 1
17/02/02 21:53:03 INFO contentpump.DelimitedTextInputFormat: 138 DelimitedSplits generated
17/02/02 21:53:04 INFO contentpump.LocalJobRunner:  completed 0%
17/02/02 21:53:04 INFO contentpump.LocalJobRunner: com.marklogic.mapreduce.MarkLogicCounter: 
17/02/02 21:53:04 INFO contentpump.LocalJobRunner: INPUT_RECORDS: 899
17/02/02 21:53:04 INFO contentpump.LocalJobRunner: OUTPUT_RECORDS: 899
17/02/02 21:53:04 INFO contentpump.LocalJobRunner: OUTPUT_RECORDS_COMMITTED: 899
17/02/02 21:53:04 INFO contentpump.LocalJobRunner: OUTPUT_RECORDS_FAILED: 0
17/02/02 21:53:04 INFO contentpump.LocalJobRunner: Total execution time: 1 sec
```
From debugging, it looks like the problem comes up when the split is at the beginning of a record.  Changing the split max size up or down changes the number of records ingested.

The reason I was testing with Japanese text is because it looks like the file seek in SplitDelimitedTextReader is byte-oriented and I wondered if it could land in the middle of a multi-byte UTF-8 encoding of a character.  Or, even if there might be a problem with \r\n on Windows systems.
[bug.zip](https://github.com/marklogic/marklogic-contentpump/files/749488/bug.zip)

**Comment by mattsunsjf** [comment]
I think this is related to bug:42027

**Comment by mattsunsjf** [comment]
These documents are missing
```
/test/00000000000798.xml
/test/00000000000799.xml
/test/00000000000800.xml
/test/00000000000801.xml
/test/00000000000802.xml
/test/00000000000803.xml
/test/00000000000804.xml
/test/00000000000805.xml
/test/00000000000806.xml
/test/00000000000807.xml
/test/00000000000808.xml
/test/00000000000809.xml
```

**Comment by mattsunsjf** [comment]
with fix cc8f247 mlcp will no longer skip a whole split if parsing the first (partial) row throws 'invalid char' exception. There is still one record missing, which is by the same cause of bug:42027: when we calculate how many bytes have been read within a DelimitedTextSplit, we didn't calculate it accurately. The CSV record returned by CSVParser has been trimmed and/or encapsulator removed. Currently we don't have access to enough information to do that accurately. I've opened an issue in Commons CSV to suggest a fix. https://issues.apache.org/jira/browse/CSV-196
